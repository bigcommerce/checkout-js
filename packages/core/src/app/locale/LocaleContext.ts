import { CurrencyService, LanguageService } from '@bigcommerce/checkout-sdk';
import { createContext } from 'react';

export interface LocaleContextType {
    language: LanguageService;
    date?: {
        inputFormat: string;
    };
    currency?: CurrencyService;
}

const LocaleContext = createContext<LocaleContextType | undefined>(undefined);

export default LocaleContext;
