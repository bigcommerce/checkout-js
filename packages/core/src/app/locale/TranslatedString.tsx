import React, { Fragment, FunctionComponent } from 'react';

import withLanguage, { WithLanguageProps } from './withLanguage';

export interface TranslatedStringProps {
    id: string;
    data?: any;
}

const TranslatedString: FunctionComponent<TranslatedStringProps & WithLanguageProps> = ({
    data,
    id,
    language,
}) => (
    <Fragment>
        { language.translate(id, data) }
    </Fragment>
);

export default withLanguage(TranslatedString);
