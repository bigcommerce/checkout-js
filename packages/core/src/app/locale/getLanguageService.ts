import { createLanguageService, LanguageConfig, LanguageService } from '@bigcommerce/checkout-sdk';

import { FALLBACK_LOCALE, FALLBACK_TRANSLATIONS } from './translations';

let languageService: LanguageService | undefined;

export default function getLanguageService(): LanguageService {
    languageService =
        languageService ??
        createLanguageService({
            fallbackLocale: FALLBACK_LOCALE,
            fallbackTranslations: FALLBACK_TRANSLATIONS,
        });

    return languageService;
}

export type InitializeLanguageService = typeof initializeLanguageService;

export function initializeLanguageService(config: LanguageConfig): LanguageService {
    languageService = createLanguageService({
        ...config,
        defaultLocale: config.locale,
        defaultTranslations: config.defaultTranslations,
        fallbackLocale: FALLBACK_LOCALE,
        fallbackTranslations: FALLBACK_TRANSLATIONS,
    });

    return languageService;
}
