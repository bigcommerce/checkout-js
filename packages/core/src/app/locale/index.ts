export { default as LocaleContext, LocaleContextType } from './LocaleContext';
export { default as createLocaleContext } from './createLocaleContext';
export { default as getDefaultTranslations } from './getDefaultTranslations';
export { default as getLanguageService, InitializeLanguageService } from './getLanguageService';
export { default as masterpassFormatLocale } from './masterpassFormatLocale';
export { default as withCurrency, WithCurrencyProps } from './withCurrency';
export { default as withLanguage, WithLanguageProps } from './withLanguage';
export { default as withDate, WithDateProps } from './withDate';
export { default as LanguageWindow, isLanguageWindow } from './LanguageWindow';
export { default as LocaleProvider } from './LocaleProvider';
export { default as TranslatedHtml, TranslatedHtmlProps } from './TranslatedHtml';
export { default as TranslatedLink, TranslatedLinkProps } from './TranslatedLink';
export { default as TranslatedString, TranslatedStringProps } from './TranslatedString';
