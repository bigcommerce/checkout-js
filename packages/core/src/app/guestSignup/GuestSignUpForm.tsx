import { withFormik, FormikProps } from 'formik';
import React, { memo, FunctionComponent } from 'react';
import { object, ref, string } from 'yup';

import { PasswordRequirements } from '../customer';
import { withLanguage, TranslatedString, WithLanguageProps } from '../locale';
import { Button, ButtonVariant } from '../ui/button';
import { Fieldset, Form, Legend } from '../ui/form';

import SignUpPasswordField from './SignUpPasswordField';

export interface SignUpFormProps {
    isSigningUp?: boolean;
    customerCanBeCreated: boolean;
    passwordRequirements: PasswordRequirements;
    onSignUp(data: SignUpFormValues): void;
}

export interface SignUpFormValues {
    password: string;
    confirmPassword: string;
}

const GuestSignUpForm: FunctionComponent<SignUpFormProps & WithLanguageProps & FormikProps<SignUpFormValues>> = ({
    isSigningUp,
    customerCanBeCreated,
    passwordRequirements: { minLength },
}) => (
    <Fieldset
        legend={
            <Legend>
                <TranslatedString id={ customerCanBeCreated ? 'customer.create_account_text' : 'customer.set_password_text' } />
            </Legend>
        }
    >
        { !customerCanBeCreated &&
            <p>
                <TranslatedString id="customer.account_created_text" />
            </p> }

        <Form className="guest-signup form">
            <SignUpPasswordField minLength={ minLength } />

            <div className="form-actions">
                <Button
                    id="createAccountButton"
                    isLoading={ isSigningUp }
                    type="submit"
                    variant={ ButtonVariant.Primary }
                >
                    <TranslatedString id={ customerCanBeCreated ? 'customer.create_account_action' : 'customer.set_password_action' } />
                </Button>
            </div>
        </Form>
    </Fieldset>
);

export default withLanguage(withFormik<SignUpFormProps & WithLanguageProps, SignUpFormValues>({
    mapPropsToValues: () => ({
        password: '',
        confirmPassword: '',
    }),
    handleSubmit: (values, { props: { onSignUp } }) => {
        onSignUp(values);
    },
    validationSchema: ({
        language,
        passwordRequirements: { description, numeric, alpha, minLength },
    }: SignUpFormProps & WithLanguageProps) => object({
        password: string()
            .required(description || language.translate('customer.password_required_error'))
            .matches(numeric, description || language.translate('customer.password_number_required_error'))
            .matches(alpha, description || language.translate('customer.password_letter_required_error'))
            .min(minLength, description || language.translate('customer.password_under_minimum_length_error'))
            .max(100, language.translate('customer.password_over_maximum_length_error')),
        confirmPassword: string()
            .required(language.translate('customer.password_confirmation_required_error'))
            .oneOf([ref('password')], language.translate('customer.password_confirmation_error')),
    }),
})(memo(GuestSignUpForm)));
