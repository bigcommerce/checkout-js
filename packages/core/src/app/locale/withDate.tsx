import { createInjectHoc, InjectHoc } from '../common/hoc';

import LocaleContext from './LocaleContext';

export interface WithDateProps {
    date: {
        inputFormat: string;
    };
}

const withDate: InjectHoc<WithDateProps> = createInjectHoc(LocaleContext, {
    displayNamePrefix: 'withDate',
    pickProps: (value, key) => key === 'date' && !!value,
});

export default withDate;
