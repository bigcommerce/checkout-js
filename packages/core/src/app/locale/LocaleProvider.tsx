import { CheckoutService, createCurrencyService, StoreConfig } from '@bigcommerce/checkout-sdk';
import { memoizeOne } from '@bigcommerce/memoize';
import React, { Component, ReactNode } from 'react';

import getLanguageService from './getLanguageService';
import LocaleContext from './LocaleContext';

export interface LocaleProviderProps {
    checkoutService: CheckoutService;
}

export interface LocaleProviderState {
    config?: StoreConfig;
}

class LocaleProvider extends Component<LocaleProviderProps> {
    state: Readonly<LocaleProviderState> = {};

    private languageService = getLanguageService();
    private unsubscribe?: () => void;

    private getContextValue = memoizeOne((config?: StoreConfig) => {
        return {
            currency: config ? createCurrencyService(config) : undefined,
            date: config
                ? {
                      inputFormat: config.inputDateFormat,
                  }
                : undefined,
            language: this.languageService,
        };
    });

    componentDidMount(): void {
        const { checkoutService } = this.props;

        this.unsubscribe = checkoutService.subscribe(
            ({ data }) => {
                this.setState({ config: data.getConfig() });
            },
            ({ data }) => data.getConfig(),
        );
    }

    componentWillUnmount(): void {
        if (this.unsubscribe) {
            this.unsubscribe();
            this.unsubscribe = undefined;
        }
    }

    render(): ReactNode {
        const { children } = this.props;
        const { config } = this.state;

        return (
            <LocaleContext.Provider value={this.getContextValue(config)}>
                {children}
            </LocaleContext.Provider>
        );
    }
}

export default LocaleProvider;
