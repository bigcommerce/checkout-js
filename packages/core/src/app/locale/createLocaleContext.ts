import { createCurrencyService, StoreConfig } from '@bigcommerce/checkout-sdk';

import getLanguageService from './getLanguageService';
import { LocaleContextType } from './LocaleContext';

export default function createLocaleContext(config: StoreConfig): Required<LocaleContextType> {
    if (!config) {
        throw new Error('Missing configuration data');
    }

    const { inputDateFormat } = config;

    return {
        currency: createCurrencyService(config),
        date: {
            inputFormat: inputDateFormat,
        },
        language: getLanguageService(),
    };
}
