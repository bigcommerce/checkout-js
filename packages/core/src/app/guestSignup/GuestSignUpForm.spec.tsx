import { mount, ReactWrapper, render } from 'enzyme';
import { noop } from 'lodash';
import React from 'react';

import { getStoreConfig } from '../config/config.mock';
import { createLocaleContext, LocaleContext, LocaleContextType } from '../locale';

import GuestSignUpForm from './GuestSignUpForm';

describe('GuestSignUpForm', () => {
    let localeContext: LocaleContextType;
    let component: ReactWrapper;
    const handleSignUp = jest.fn();
    const passwordRequirements = {
        minLength: 5,
        description: 'Password should be ...',
        alpha: /w/,
        numeric: /d/,
    };

    beforeEach(() => {
        localeContext = createLocaleContext(getStoreConfig());

        component = mount(
            <LocaleContext.Provider value={localeContext}>
                <GuestSignUpForm
                    customerCanBeCreated={true}
                    onSignUp={handleSignUp}
                    passwordRequirements={passwordRequirements}
                />
            </LocaleContext.Provider>,
        );
    });

    it('matches snapshot', () => {
        const shallowComponent = render(
            <LocaleContext.Provider value={localeContext}>
                <GuestSignUpForm
                    customerCanBeCreated={true}
                    onSignUp={noop}
                    passwordRequirements={passwordRequirements}
                />
            </LocaleContext.Provider>,
        );

        expect(shallowComponent).toMatchSnapshot();
    });

    it('matches snapshot when customer cannot be created', () => {
        const shallowComponent = render(
            <LocaleContext.Provider value={localeContext}>
                <GuestSignUpForm
                    customerCanBeCreated={false}
                    onSignUp={noop}
                    passwordRequirements={passwordRequirements}
                />
            </LocaleContext.Provider>,
        );

        expect(shallowComponent).toMatchSnapshot();
    });

    it('notifies when user clicks on "create account" button', async () => {
        component
            .find('input[name="password"]')
            .simulate('change', { target: { value: 'password1', name: 'password' } });

        component
            .find('input[name="confirmPassword"]')
            .simulate('change', { target: { value: 'password1', name: 'confirmPassword' } });

        component.find('form').simulate('submit');

        await new Promise((resolve) => process.nextTick(resolve));

        expect(handleSignUp).toHaveBeenCalled();
    });

    it('displays error message if password is not valid', async () => {
        component
            .find('input[name="password"]')
            .simulate('change', { target: { value: '1', name: 'password' } });

        component.find('form').simulate('submit');

        await new Promise((resolve) => process.nextTick(resolve));

        component.update();

        expect(component.find('[data-test="password-field-error-message"]').text()).toEqual(
            passwordRequirements.description,
        );
    });

    it('displays error message if passwords are missing', async () => {
        component.find('form').simulate('submit');

        await new Promise((resolve) => process.nextTick(resolve));

        component.update();

        expect(component.find('[data-test="password-field-error-message"]').text()).toEqual(
            passwordRequirements.description,
        );

        expect(component.find('[data-test="confirm-password-field-error-message"]').text()).toBe(
            'This field is required',
        );
    });

    it('displays error message if passwords do not match', async () => {
        component
            .find('input[name="password"]')
            .simulate('change', { target: { value: 'password1', name: 'password' } });

        component
            .find('input[name="confirmPassword"]')
            .simulate('change', { target: { value: 'password2', name: 'confirmPassword' } });

        component.find('form').simulate('submit');

        await new Promise((resolve) => process.nextTick(resolve));

        component.update();

        expect(component.find('[data-test="confirm-password-field-error-message"]').text()).toBe(
            'Passwords do not match',
        );
    });
});
