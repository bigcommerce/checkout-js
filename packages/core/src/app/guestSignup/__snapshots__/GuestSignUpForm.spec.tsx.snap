// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GuestSignUpForm matches snapshot 1`] = `
<fieldset
  class="form-fieldset"
>
  <legend
    class="form-legend optimizedCheckout-headingSecondary"
  >
    Create an account for a faster checkout in the future
  </legend>
  <div
    class="form-body"
  >
    <div>
      <form
        class="guest-signup form"
        novalidate=""
      >
        <div
          class="form-field"
        >
          <label
            class="form-label optimizedCheckout-form-label"
            for="password"
          >
            Password 
            <small>
              5-character minimum, case sensitive
            </small>
          </label>
          <input
            class="form-input optimizedCheckout-form-input"
            id="password"
            name="password"
            type="password"
            value=""
          />
        </div>
        <div
          class="form-field"
        >
          <label
            class="form-label optimizedCheckout-form-label"
            for="confirmPassword"
          >
            Confirm Password
          </label>
          <input
            class="form-input optimizedCheckout-form-input"
            id="confirmPassword"
            name="confirmPassword"
            type="password"
            value=""
          />
        </div>
        <div
          class="form-actions"
        >
          <button
            class="button button--primary optimizedCheckout-buttonPrimary"
            id="createAccountButton"
            type="submit"
          >
            Create Account
          </button>
        </div>
      </form>
    </div>
  </div>
</fieldset>
`;

exports[`GuestSignUpForm matches snapshot when customer cannot be created 1`] = `
<fieldset
  class="form-fieldset"
>
  <legend
    class="form-legend optimizedCheckout-headingSecondary"
  >
    Setup a password for a faster checkout in the future
  </legend>
  <div
    class="form-body"
  >
    <p>
      An account has been created for you, but you'll need to setup a password to sign in.
    </p>
    <div>
      <form
        class="guest-signup form"
        novalidate=""
      >
        <div
          class="form-field"
        >
          <label
            class="form-label optimizedCheckout-form-label"
            for="password"
          >
            Password 
            <small>
              5-character minimum, case sensitive
            </small>
          </label>
          <input
            class="form-input optimizedCheckout-form-input"
            id="password"
            name="password"
            type="password"
            value=""
          />
        </div>
        <div
          class="form-field"
        >
          <label
            class="form-label optimizedCheckout-form-label"
            for="confirmPassword"
          >
            Confirm Password
          </label>
          <input
            class="form-input optimizedCheckout-form-input"
            id="confirmPassword"
            name="confirmPassword"
            type="password"
            value=""
          />
        </div>
        <div
          class="form-actions"
        >
          <button
            class="button button--primary optimizedCheckout-buttonPrimary"
            id="createAccountButton"
            type="submit"
          >
            Save Password
          </button>
        </div>
      </form>
    </div>
  </div>
</fieldset>
`;
